src/tex/optimize.cpp

```diff
@@ -157,14 +158,18 @@ auto compute_optimization_steps(const Texture &file, const Settings &sets) noexc
             res.add_transparent_alpha = true;
 
     const bool opt_mip = optimal_mip_count(file.get_dimension()) == info.mipLevels;
-    if (sets.mipmaps && (!opt_mip || res.resize)) // resize removes mips
+    if ((sets.mipmaps && !opt_mip) || (info.mipLevels > 1 && res.resize)) // resize removes mips if there are any, regenerate them
         res.mipmaps = true;
 
     // I prefer to keep steps independent, but this one has to depend on add_transparent_alpha. If we add an alpha, the output format must have alpha
     res.best_format = best_output_format(file,
                                          sets,
                                          res.add_transparent_alpha);
 
+    // Ensure result is saved without alpha in cases where it could be problematic (i.e. resizing).
+    if (res.resize && res.best_format == sets.output_format.uncompressed_without_alpha)
+        res.convert = true;
```

Comment:
You are only checking uncompressed. I think it also applies to compressed?

Reply:
We should convert/compress on every resize, yes, because for compressed textures resizing requires decompressing. But this is implicit in the `optimize` function already (`must_decompress` flag), even if `res.convert == false`.

This is why I am thinking about refactoring the whole steps -> optimize logic a bit. They kinda clash into each other with their responsibilites. I'm just not sure how I would do it (for example the decompression requirement - should it be part of the optimization steps or the optimize function? I think both are justifiable).

Reply:
Please add a comment explaining this

The reason for `compute_optimization_steps` to be separated is to allow the caller to change the steps and to allow dry run. CAO5 logic was a mess, so this is what I came up with. However, I agree that it will get hard to follow the more steps we add

Reply:
Yeah, it's good to have it separated especially if a fine-grain per-file control will be added one day (as we discussed on Discord).

I actually went back and rechecked everything and decided that this conversion should happen implicitely. Also, I'm dumb and didn't know all transformations including mipmap generation requires decompressed texture, so yeah, fixed that.

Added a few more tests to be sure this time.